### PR TITLE
drivers: display: Kyocera memory-in-pixel display

### PR DESCRIPTION
**Pull request by alen-daniel**
shields: display: added display shield for Kyocera TN0216ANVNANN

Added a shield and display driver for use with the Kyocera TN0216ANVNANN SPI display driver

Added kyocera memory-in-pixel LCD TN0216ANVNANN shield

* TN0103ANVNANN-GN00
* TN0181ANVNANN-GN00
* TN0216ANVNANN-GN00 <- Tested
* TN0227ANVNANN-GNX03

The changes in this PR were modelled after this PR, thank you to @rgujju 
https://github.com/zephyrproject-rtos/zephyr/pull/30299

The following devicetree configuration is an example:
```
&spi1 {
	compatible = "nordic,nrf-spi";
	status = "okay";
	pinctrl-0 = <&spi1_default>;
	pinctrl-1 = <&spi1_sleep>;
	pinctrl-names = "default", "sleep";
	cs-gpios = <&gpio0 14 GPIO_ACTIVE_HIGH>;
	tn0xxx: tn0xxx@0 {
		compatible = "kyo,tn0xxx";
		status = "okay";
		reg = <0>;
		width = <320>;
		height = <176>;
		spi-max-frequency = <500000>;
	};
};
```

The following configuration works for me: 
```
# display
CONFIG_DISPLAY=y 
CONFIG_SPI=y
CONFIG_TN0XXX=y
# LVGL
CONFIG_LVGL=y
CONFIG_LV_Z_BITS_PER_PIXEL=1
CONFIG_LV_Z_VDB_SIZE=100
CONFIG_LV_MEM_CUSTOM=y
CONFIG_LV_Z_MEM_POOL_NUMBER_BLOCKS=2
CONFIG_LV_COLOR_DEPTH_1=y
```

### CODE REVIEW

File: samples/drivers/display/sample.yaml [hunk 1]
```diff
@@ -79,6 +79,15 @@ tests:
     harness: console
     harness_config:
       fixture: fixture_display
+  sample.display.tn0216anvnann:
+    platform_allow: nrf52dk_nrf52833
```
Comment (kartben) [inline samples/drivers/display/sample.yaml:83]: I think you want



```suggestion
    platform_allow: nrf52dk_nrf52832
```

Comment (alen-daniel) [reply samples/drivers/display/sample.yaml]: thanks for pointing that out! fixed and pushed a commit

**Comment by alen-daniel** [comment]
> @alen-daniel btw as you start amending your initial PR based on feedback from CI etc you will want to keep squashing your changes in your initial commit, and then force push the changes to your branch. This way at the end of the day when the PR gets merged all your changes are reflected in the same meaningful commit.

Will do! I intended to do that as I make my way through all the incoming PR feedback.

For now I am moving this PR to a draft stage while I work through the CI failures.

File: drivers/display/display_tn0xxx.c [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+	lv_disp_t *disp = lv_disp_get_default();
+	struct lvgl_disp_data *disp_data = disp->driver->user_data;
+	struct display_capabilities *caps = &disp_data->cap;
+	struct tn0xxx_data_s *data = dev->data;
+
+	if (pf & caps->supported_pixel_formats) {
+		caps->current_pixel_format = pf;
+		data->pixel_format = pf;
+		lv_disp_drv_update(disp, disp->driver);
+		return 0;
```
Comment (jfischer-no) [inline drivers/display/display_tn0xxx.c:109]: The driver may only use the API provided by https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/display.h. The driver may not call any gui library or subsystem, and certainly not any third party library or subsystem.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: drivers/display/display_tn0xxx.c (hunk 2 above)
Comment (alen-daniel) [reply drivers/display/display_tn0xxx.c]: Thank you for this feedback, good points. I have updated the commit to reflect this

File: drivers/display/display_tn0xxx.c [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+			single_line_buffer[buff_index++] = bitmap_buffer[bitmap_buffer_index++];
+		}
+		for (int i = 0; i < LCD_DUMMY_SPI_CYCLES_LEN_BITS / TN0XXX_PIXELS_PER_BYTE; i++) {
+			single_line_buffer[buff_index++] = DUMMY_BYTE;
+		}
+
+		int len = sizeof(single_line_buffer);
+
+		struct spi_buf tx_buf = {.buf = single_line_buffer, .len = len};
+		struct spi_buf_set tx_bufs = {.buffers = &tx_buf, .count = 1};
```
[the remark this answers is not in the document]
Comment (alen-daniel) [reply drivers/display/display_tn0xxx.c]: Done

File: drivers/display/display_tn0xxx.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+		caps->supported_pixel_formats);
+	return -ENOTSUP;
+}
+
+static int update_display(const struct device *dev, uint16_t start_line, uint16_t num_lines,
+			  const uint8_t *bitmap_buffer)
+{
+
+	const struct tn0xxx_config_s *config = dev->config;
+	uint8_t single_line_buffer[(TN0XXX_PANEL_WIDTH + LCD_DUMMY_SPI_CYCLES_LEN_BITS +
```
[the remark this answers is not in the document]
Comment (alen-daniel) [reply drivers/display/display_tn0xxx.c]: Agreed, done.

**Comment by alen-daniel** [comment]
@erwango just wanted to nudge you about this PR

@jfischer-no hey I have updated the PR with your requested changes, pls have a look 🙇

File: boards/shields/tn0xxx/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
+.. _tn0xxx_shield:
```
Comment (jfischer-no) [inline boards/shields/tn0xxx/doc/index.rst:1]: Please split it into two commits, one for the driver and bindings, and one to add the shield.

File: boards/shields/tn0xxx/doc/index.rst [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+| RST           | Reset                                                   |
++---------------+---------------------------------------------------------+
+| VCOM          | VCOM Inversion Polarity Input (VCOM can be controlled   |
+|               | through SW)                                             |
++---------------+---------------------------------------------------------+
+
+The zephyr display driver for TN0XXX only controls the SPI lines (SCS, SI, SCLK) - the power up and power down sequence is delegated to the user. This is because the datasheet specifies a power up and down sequence with timings specified between toggling of RST, VCOM and input SPI lines. This MIP display is typically used in low power applications, and as such having user control over the power down (or sleep) and power up (or wake up) options are important. Since the API to power on and off the device is not present in the zephyr driver display API, the control of these GPIOs per the datasheet is delegated to the user. Also note: VCOM = “L” is necessary when RST = ”H” otherwise the current consumption increases by shoot-through-current in panel (several mA).
```
Comment (jfischer-no) [inline boards/shields/tn0xxx/doc/index.rst:31]: This line is to long, max 100 characters

File: drivers/display/display_tn0xxx.c [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+static int tn0xxx_set_pixel_format(const struct device *dev, const enum display_pixel_format pf)
+{
+	if (pf == PIXEL_FORMAT_MONO01) {
+		return 0;
+	}
+	LOG_ERR("not supported");
+	return -ENOTSUP;
+}
```
[the remark this answers is not in the document]
Comment (alen-daniel) [reply drivers/display/display_tn0xxx.c]: done! thank you for the feedback

File: boards/shields/tn0xxx/doc/index.rst (hunk 6 above)
Comment (alen-daniel) [reply boards/shields/tn0xxx/doc/index.rst]: done! thank you for the feedback

File: boards/shields/tn0xxx/doc/index.rst (hunk 5 above)
Comment (alen-daniel) [reply boards/shields/tn0xxx/doc/index.rst:1]: done! thank you for the feedback

**Pull request closed without merging**